### DISCUSSION TITLE
Documentation issues

### DISCUSSION THREAD

**Discussion by czkz** [vlang/v#12687] [Ideas]
tldr: Issues in doc/docs.md from a newcomer's PoV

I wrote this a month ago, while reading the docs, with no prior knowledge of V, Go, Rust, Nim, in hopes that it will provide a point of view experienced V developers don't have.
I reviewed the text today and removed all the issues that got fixed, so all the remaining issues/links are still valid (right now 3af53e2 is the latest commit).
_Italic text_ is used for remarks I made today, **with hindsight of using the language for a month**.

There were also a lot of typos, which I separated into #12227.

There are probably better ways to format this, but I'm not aware of any :confused:

---

---

https://github.com/vlang/v/blob/master/doc/docs.md#strings
Missing info on how to convert []byte back to string

_Found this through trial and error, not sure it it's the intended way:_
```v
b := 'abc'.bytes()
s := string(b)
```

---

https://github.com/vlang/v/blob/master/doc/docs.md#fixed-size-arrays
> anums := fnums[0..fnums.len]

fnums[..] is much more compact, and works with anonymous variables (doesn't require a variable to get its .len).
Not sure if this syntax is a bug, but it works with V 0.2.4 5010d18.

_Still works in V 0.2.4 a006090_

---

https://github.com/vlang/v/blob/master/doc/docs.md#fixed-size-arrays
Missing info on how to convert a dynamic array to a fixed array

_Still haven't found a way to do it, maybe I should make a Feature Request? Fixed arrays are very hard to use without it._

---

https://github.com/vlang/v/blob/master/doc/docs.md#type-checks-and-casts
> Mutable variables can change, and doing a cast would be unsafe

DOESN'T MAKE SENSE

_It's not explained why doing a cast would be unsafe, and it's not something obvious; I really don't know why it would be unsafe_

---

https://github.com/vlang/v/blob/master/doc/docs.md#short-struct-literal-syntax
> you can omit the struct name when it's already known

But nothing is omitted

_Still can't figure out what was supposed to be omitted_

---

https://github.com/vlang/v/blob/master/doc/docs.md#mutable-arguments
> Use user.register() or user = register(user) instead of register(mut user).

Misleading into thinking that these are the ways the code above can be used

---

https://github.com/vlang/v/blob/master/doc/docs.md#manage-packages
> Read more:

Missing link

_I now see that it's not a missing link, but refers to "Briefly:" above. This should probably be removed/reworded._

---

https://github.com/vlang/v/blob/master/doc/docs.md#module-options
> Or you can update all your modules:

No info about upgrade command, how it's different from update with no options

---

https://github.com/vlang/v/blob/master/doc/docs.md#publish-package
It's strange that there's language info, then package management, then language info again. Consider relocating the "Modules" section.

---

https://github.com/vlang/v/blob/master/doc/docs.md#interfaces
Would be nice to have questionable design choices explained - no "implement" keyword seems like a bad design choice to me (wouldn't it be tidious to implement large interfaces when you have to manually remember all the fields you have to define?)
If I knew why it was made, or maybe there was some counterpoint to what I thought, it would've made sense but without any of that people might tend to think that it's just objectively bad.

---

https://github.com/vlang/v/blob/master/doc/docs.md#channel-select
> `else` and `> timeout` are mutually exclusive

'>' is a typo or it's unclear what it means

---

https://github.com/vlang/v/blob/master/doc/docs.md#channel-select
This is the first snipped I tried to compile and it failed with "library 'atomic' not found". I assumed it just wasn't implemented yet, but that would be nice to know beforehand, for example in the beginning of the docs.

_Two weeks later I discovered that it actually was implemented, but I had to install libatomic-devel. Which is a huge problem:_
- _The language is very alpha, some features are not implemented now, but some of them are (or might be) in the docs._
- _And it's advertized that V requires (or maybe will require?) no dependencies._
- _And it fails with a very cryptic message (Invalid relocation entry '.rela.debug_info' @ 000004f6...)._
- _So most people will probably think that the language is broken, that docs describe what the language wants to be and not what it is._
- _And, thinking that all those awesome features are not real, they will likely leave._
- _Which is especially sad because the language does have all these features._

_Later I discovered that there were some other libs I had to install in order for V to work; I don't remember them now, but it's a good idea to have a list of things required for V to work at least with built-in features like channels._

---

https://github.com/vlang/v/blob/master/doc/docs.md#memory-management
> Remaining small percentage of objects is freed via reference counting.

But on https://vlang.io it's said: "Remaining small percentage of objects is freed via GC".

_I don't know which one is correct so I can't fix it._

---

https://github.com/vlang/v/blob/master/doc/docs.md#structs-with-reference-fields
> for now data structures <...> understanding that it is unsafe

malformed sentence

_Also it's not explained why an unsafe operation is not wrapped in an unsafe{} block (maybe it should be?)._

---

https://github.com/vlang/v/blob/master/doc/docs.md#embed_file
> Instead, it will be loaded the first time your program calls `f.data()` at runtime...

What is `f.data()`?
The code snippet above doesn't have `f.data()`, nor there are any explanations later in the docs.

---

https://github.com/vlang/v/blob/master/doc/docs.md#compile-time-reflection
Not clear whether get_string and get_int are built-in functions or example functions.
Add something along the lines of
```v
fn get_string(data string, field_name string) string {
    // ...
}
fn get_int(data string, field_name string) string {
    // ...
}
```

_There are probably no global built-in functions in V (except `println()`, etc.), but a newcomer wouldn't know that._

---

https://github.com/vlang/v/blob/master/doc/docs.md#limited-operator-overloading
> Assignment operators <...> are auto generated <...> though they must return the same type.

What does "though they must return the same type" mean - aren't they auto generated?

---

https://github.com/vlang/v/blob/master/doc/docs.md#cross-platform-shell-scripts-in-v
> if files.len > 0 {

Seems redundant, the foreach loop wouldn't run with zero elements anyway

**Comment by Cons-Cat** [comment]
> Missing info on how to convert a dynamic array to a fixed array
>
> Still haven't found a way to do it, maybe I should make a Feature Request? Fixed arrays are very hard to use without it.

I asked about this once: #11063

**Comment by JalonSolov** [comment]
> https://github.com/vlang/v/blob/master/doc/docs.md#strings Missing info on how to convert []byte back to string
> 
> _Found this through trial and error, not sure it it's the intended way:_
> 
> ```v
> b := 'abc'.bytes()
> s := string(b)
> ```

That's the simplest way, and yes it's intended.You are casting the array of bytes to a string.

> https://github.com/vlang/v/blob/master/doc/docs.md#fixed-size-arrays
> 
> > anums := fnums[0..fnums.len]
> 
> fnums[..] is much more compact, and works with anonymous variables (doesn't require a variable to get its .len). Not sure if this syntax is a bug, but it works with V 0.2.4 [5010d18](https://github.com/vlang/v/commit/5010d18f27e1ee2dc4cc82a5cab9ffca40d106d6).

The 2nd form is the shorthand for the first.  Both are correct.

> _Still works in V 0.2.4 [a006090](https://github.com/vlang/v/commit/a006090b0884c111e09f8321dc41b9aaa0a6d649)_
> 
> https://github.com/vlang/v/blob/master/doc/docs.md#fixed-size-arrays Missing info on how to convert a dynamic array to a fixed array
> 
> _Still haven't found a way to do it, maybe I should make a Feature Request? Fixed arrays are very hard to use without it._

Why?  Fixed arrays exist for a very specific purpose - to have a guaranteed space that won't get reallocated, and allocated on the stack for speed.  Unless you need those specific features, don't use them.

However, if you really need it, the only way to convert between them is with a for loop.
```v
f := [0, 1, 2]!
mut d := []int{}

for i in f {
        d << i
}

println('$f, $d')
```

> https://github.com/vlang/v/blob/master/doc/docs.md#type-checks-and-casts
> 
> > Mutable variables can change, and doing a cast would be unsafe
> 
> DOESN'T MAKE SENSE
> 
> _It's not explained why doing a cast would be unsafe, and it's not something obvious; I really don't know why it would be unsafe_

It does make sense.  It is talking about casting to a Sum Type, and casting to a Sum Type is unsafe, as a Sum Type could be any of the types, and you might be casting to the wrong one.  Or it may even have changed since you last looked at it.  If you're _sure_ you know what's going to happen, then you can mark it as mutable so the compiler knows not to squawk about it.

> https://github.com/vlang/v/blob/master/doc/docs.md#short-struct-literal-syntax
> 
> > you can omit the struct name when it's already known
> 
> But nothing is omitted
> 
> _Still can't figure out what was supposed to be omitted_

The comment is just in the wrong place.  It's talking about the `assert` statement just after the struct definition.  Instead of having to type
```v
assert Point.p.y == 4
```
you only need
```v
assert p.y == 4
```

> https://github.com/vlang/v/blob/master/doc/docs.md#mutable-arguments
> 
> > Use user.register() or user = register(user) instead of register(mut user).
> 
> Misleading into thinking that these are the ways the code above can be used

It is referring to the prior statement `It is preferable to return values instead of modifying arguments.`

I agree that it is somewhat confusing, since only the first statement will work with the given code.

> https://github.com/vlang/v/blob/master/doc/docs.md#manage-packages
> 
> > Read more:
> 
> Missing link
> 
> _I now see that it's not a missing link, but refers to "Briefly:" above. This should probably be removed/reworded._

Agreed.

> https://github.com/vlang/v/blob/master/doc/docs.md#module-options
> 
> > Or you can update all your modules:
> 
> No info about upgrade command, how it's different from update with no options

It is documented in the `v help` output.  There has been discussion about ways to include the `v help` output with the regular docs, but that hasn't happened, yet.
```
$ v help update
Usage:
  v update [MODULE]...
        Updates each MODULE.

        With no MODULE, ALL installed modules are updated to their latest versions.

Options:
  -help        - Show usage info.
  -v           - Print more details about the update.
  -server-url  - When doing network operations, use this vpm server. Can be given multiple times.

$
```
```
$ v help upgrade
Usage: v upgrade

Upgrade all outdated modules.

Options:
  -help        - Show usage info.
  -v           - Print more details about the upgrade.

$
```

> https://github.com/vlang/v/blob/master/doc/docs.md#publish-package It's strange that there's language info, then package management, then language info again. Consider relocating the "Modules" section.

Agreed.

> https://github.com/vlang/v/blob/master/doc/docs.md#interfaces Would be nice to have questionable design choices explained - no "implement" keyword seems like a bad design choice to me (wouldn't it be tedious to implement large interfaces when you have to manually remember all the fields you have to define?) If I knew why it was made, or maybe there was some counterpoint to what I thought, it would've made sense but without any of that people might tend to think that it's just objectively bad.

I'm not sure how an "implements" keyword would help.  You still need to create all the pieces yourself, so adding "implements" is superfluous.

> https://github.com/vlang/v/blob/master/doc/docs.md#channel-select
> 
> > `else` and `> timeout` are mutually exclusive
> 
> '>' is a typo or it's unclear what it means

Typo.

> https://github.com/vlang/v/blob/master/doc/docs.md#channel-select This is the first snipped I tried to compile and it failed with "library 'atomic' not found". I assumed it just wasn't implemented yet, but that would be nice to know beforehand, for example in the beginning of the docs.
> 
> _Two weeks later I discovered that it actually was implemented, but I had to install libatomic-devel. Which is a huge problem:_
> 
>     * _The language is very alpha, some features are not implemented now, but some of them are (or might be) in the docs._
> 
>     * _And it's advertised that V requires (or maybe will require?) no dependencies._
> 
>     * _And it fails with a very cryptic message (Invalid relocation entry '.rela.debug_info' @ 000004f6...)._
> 
>     * _So most people will probably think that the language is broken, that docs describe what the language wants to be and not what it is._
> 
>     * _And, thinking that all those awesome features are not real, they will likely leave._
> 
>     * _Which is especially sad because the language does have all these features._
> 
> 
> _Later I discovered that there were some other libs I had to install in order for V to work; I don't remember them now, but it's a good idea to have a list of things required for V to work at least with built-in features like channels._

Well, none of them are necessary for base V to work.  They are needed for specific modules or features to work.  Yes, some of it could be handled better, and there have been some changes to help let you know when you need to install something extra, but that hasn't been "back-ported" to all the code since it was added later.  V is still a work-in-progress.
 
> https://github.com/vlang/v/blob/master/doc/docs.md#memory-management
> 
> > Remaining small percentage of objects is freed via reference counting.
> 
> But on https://vlang.io it's said: "Remaining small percentage of objects is freed via GC".
> 
> _I don't know which one is correct so I can't fix it._

They are both correct, as reference counting can be considered as a type of GC.

> https://github.com/vlang/v/blob/master/doc/docs.md#structs-with-reference-fields
> 
> > for now data structures <...> understanding that it is unsafe
> 
> malformed sentence
> 
> _Also it's not explained why an unsafe operation is not wrapped in an unsafe{} block (maybe it should be?)._

As it says just about that, `Zero-value references, or nil pointers, will NOT be supported in the future`.  For now you can, and you don't have to wrap them in unsafe{}, but later you won't be able to use them at all.

> https://github.com/vlang/v/blob/master/doc/docs.md#embed_file
> 
> > Instead, it will be loaded the first time your program calls `f.data()` at runtime...
> 
> What is `f.data()`? The code snippet above doesn't have `f.data()`, nor there are any explanations later in the docs.

Yeah, that needs to be fixed.  It should be `embedded_file.data()` to match the example code.

> https://github.com/vlang/v/blob/master/doc/docs.md#compile-time-reflection Not clear whether get_string and get_int are built-in functions or example functions. Add something along the lines of
> 
> ```v
> fn get_string(data string, field_name string) string {
>     // ...
> }
> fn get_int(data string, field_name string) string {
>     // ...
> }
> ```
> 
> _There are probably no global built-in functions in V (except `println()`, etc.), but a newcomer wouldn't know that._

As it says at the top of that example:
```v
// TODO: not fully implemented
```
There are quite a few builtin functions - look at [modules.vlang.io](https://modules.vlang.io) under `builtin`.

However, you are correct - they aren't global - they're in the builtin module.  It is simply auto-imported for every V program.

> https://github.com/vlang/v/blob/master/doc/docs.md#limited-operator-overloading
> 
> > Assignment operators <...> are auto generated <...> though they must return the same type.
> 
> What does "though they must return the same type" mean - aren't they auto generated?

If you write
```v
a := int(1)
b := f64(3.5)
```
then `a += b` won't work, since `a` and `b` aren't the same type.

It only autogenerates the overloads for things that are the same type.

> https://github.com/vlang/v/blob/master/doc/docs.md#cross-platform-shell-scripts-in-v
> 
> > if files.len > 0 {
> 
> Seems redundant, the foreach loop wouldn't run with zero elements anyway

It's a commented out example, but it is also a safeguard against `files` not having any elements... no reason to even try the `for` loop, then.

**Comment by Cons-Cat** [comment]
> They are both correct, as reference counting can be considered as a type of GC.

I have often seen it said that Swift has no garbage collection. Whatever the precise definition of GC actually is, I think people read GC and have a very different idea of what algorithms that might mean than when they read automatic reference counting.

**Comment by JalonSolov** [comment]
That is often the case, and has lead to some rather heated debates.

It all comes down to the program cleaning up unused bits & bobs at runtime, no matter how that is achieved.

If it didn't have _something_ like this, you would have to manually add `free()` calls, and nobody really wants to do that.

**Comment by czkz** [comment]
> > Missing info on how to convert a dynamic array to a fixed array
> > Still haven't found a way to do it, maybe I should make a Feature Request? Fixed arrays are very hard to use without it.
> 
> I asked about this once: #11063

The `[1, 2, 3]!` syntax seems to work only with array literals; consider this code:
```v
import os

fn foo(data [32]byte) {
}

fn main() {
	data := os.args[0].bytes()
	// foo(???)
}
```

**Reply by ntrel** [reply]
You can use `mut arr := [32]byte {}` to create a default initialized array. Then you can copy the bytes manually. Or you may be able to do `*(&byte[32](data.data))` but I haven't tried it.

**Comment by czkz** [comment]
@JalonSolov Thanks for an extensive review! It explained most of the things to me, and I'll make a pull request soon to make them clear for others.

There are a still few things I'm not sure about though.

---

> > https://github.com/vlang/v/blob/master/doc/docs.md#short-struct-literal-syntax
> > > you can omit the struct name when it's already known
> > 
> > 
> > But nothing is omitted
> > _Still can't figure out what was supposed to be omitted_
> 
> The comment is just in the wrong place. It's talking about the `assert` statement just after the struct definition. Instead of having to type
> 
> ```v
> assert Point.p.y == 4
> ```
> 
> you only need
> 
> ```v
> assert p.y == 4
> ```

I'm not aware of any languages that use this syntax. Should I just remove the comment?

---

> > https://github.com/vlang/v/blob/master/doc/docs.md#module-options
> > > Or you can update all your modules:
> > 
> > 
> > No info about upgrade command, how it's different from update with no options
> 
> It is documented in the `v help` output. There has been discussion about ways to include the `v help` output with the regular docs, but that hasn't happened, yet.

- `v help update` says "ALL installed modules are updated to their latest versions".
- `v help upgrade` says "Upgrade all outdated modules".
- `v help outdated` says "List all installed modules that need updates".

So, `update` updates all modules, `upgrade` updates all outdated modules, which are, according to `outdated`, "modules that need updates".

So `update` and `upgrade` do the same thing, right?

---

> > https://github.com/vlang/v/blob/master/doc/docs.md#interfaces Would be nice to have questionable design choices explained - no "implement" keyword seems like a bad design choice to me (wouldn't it be tedious to implement large interfaces when you have to manually remember all the fields you have to define?) If I knew why it was made, or maybe there was some counterpoint to what I thought, it would've made sense but without any of that people might tend to think that it's just objectively bad.
> 
> I'm not sure how an "implements" keyword would help. You still need to create all the pieces yourself, so adding "implements" is superfluous.

`implements` allows the compiler to warn me if I failed to implement an interface, which is helpful in multiple scenarios:

#### 1. I mixed up types
E.g. I declared `score int`, but an interface expects `score f64`

#### 2. I forgot something
E.g. while implementing `WebBrowser` I only defined 1317 out of 1318 methods it requires
#### 3. I maintain multiple implementations
E.g. I have an interface `Renderer` implemented by `Opengl` on Linux and `Directx` on Windows.
I add new features to `Renderer`, and implement them right away.
I know I implemented `Opengl` right, because it works, but I can't be certain that I didn't make mistakes `1.` and `2.` in `Directx` because I'd have to boot to Windows and run it to know for sure

This is simillar to dynamically typed languages vs statically typed:
In Javascript and Python you have to _use_ some code to know that there are no type errors, while statically typed languages catch them during compilation.
Without `implements` you have to _use_ a struct through an interface to know that it implements that interface, while with `implements` you know it during compilation.

---

> > https://github.com/vlang/v/blob/master/doc/docs.md#channel-select
> > <...> channels need libatomic <...> 
> 
> Well, none of them are necessary for base V to work. They are needed for specific modules or features to work. Yes, some of it could be handled better, and there have been some changes to help let you know when you need to install something extra, but that hasn't been "back-ported" to all the code since it was added later. V is still a work-in-progress.

V indeed works without them, and some modules have readmes with dependencies and instructions to install them (like [sqlite](https://modules.vlang.io/sqlite.html)).
But since channels are imported implicitly, maybe their dependencies should be listed with V's base dependencies, or even better, in an error message.

---

> > https://github.com/vlang/v/blob/master/doc/docs.md#memory-management
> > > Remaining small percentage of objects is freed via reference counting.
> 
> > But on https://vlang.io it's said: "Remaining small percentage of objects is freed via GC".
> > _I don't know which one is correct so I can't fix it._
> 
> They are both correct, as reference counting can be considered as a type of GC.

I assume they both should be reference counting then.

---

> > https://github.com/vlang/v/blob/master/doc/docs.md#embed_file
> > > Instead, it will be loaded the first time your program calls `f.data()` at runtime...
> > 
> > 
> > What is `f.data()`? The code snippet above doesn't have `f.data()`, nor there are any explanations later in the docs.
> 
> Yeah, that needs to be fixed. It should be `embedded_file.data()` to match the example code.

But there's `embedded_file.to_string()` in the example code. Should it be like in the example or should the example be changed?

---

> 
> > https://github.com/vlang/v/blob/master/doc/docs.md#compile-time-reflection Not clear whether get_string and get_int are built-in functions or example functions. Add something along the lines of
> > ```v
> > fn get_string(data string, field_name string) string {
> >     // ...
> > }
> > fn get_int(data string, field_name string) string {
> >     // ...
> > }
> > ```
> 
> As it says at the top of that example:
> 
> ```v
> // TODO: not fully implemented
> ```

~~This should probably be highlighted a bit~~
It looks like it's implemented now! I'll come up with a full working example and move `decode_User()` to a different block.

---

> > https://github.com/vlang/v/blob/master/doc/docs.md#cross-platform-shell-scripts-in-v
> > > if files.len > 0 {
> > 
> > 
> > Seems redundant, the foreach loop wouldn't run with zero elements anyway
> 
> It's a commented out example, but it is also a safeguard against `files` not having any elements... no reason to even try the `for` loop, then.

If `files` array was empty, the loop would run zero times - `for e in []int{} { println(e) }` produces no output, so I assume `if` is safe to remove.

**Comment by JalonSolov** [comment]
> @JalonSolov Thanks for an extensive review! It explained most of the things to me, and I'll make a pull request soon to make them clear for others.
> 
> There are a still few things I'm not sure about though.
> 
> > > https://github.com/vlang/v/blob/master/doc/docs.md#short-struct-literal-syntax
> > > > you can omit the struct name when it's already known
> > > 
> > > 
> > > But nothing is omitted
> > > _Still can't figure out what was supposed to be omitted_
> > 
> > 
> > The comment is just in the wrong place. It's talking about the `assert` statement just after the struct definition. Instead of having to type
> > ```v
> > assert Point.p.y == 4
> > ```
> >
> > you only need
> > ```v
> > assert p.y == 4
> > ```
> 
> I'm not aware of any languages that use this syntax. Should I just remove the comment?

It might be confusing to some without it.  "Why don't you have to use `Point` there?"
 
> > > https://github.com/vlang/v/blob/master/doc/docs.md#module-options
> > > > Or you can update all your modules:
> > > 
> > > 
> > > No info about upgrade command, how it's different from update with no options
> > 
> > 
> > It is documented in the `v help` output. There has been discussion about ways to include the `v help` output with the regular docs, but that hasn't happened, yet.
> 
>     * `v help update` says "ALL installed modules are updated to their latest versions".
> 
>     * `v help upgrade` says "Upgrade all outdated modules".
> 
>     * `v help outdated` says "List all installed modules that need updates".
> 
> 
> So, `update` updates all modules, `upgrade` updates all outdated modules, which are, according to `outdated`, "modules that need updates".
> 
> So `update` and `upgrade` do the same thing, right?

No... look at the `v help` output more closely (from the command line, not anything in docs.md):
```
Usage:
  v update [MODULE]...
        Updates each MODULE.
```
In other words, `v update` will (try to) update all modules _if_ you don't list any, but if you do list some, _only_ those will be updated.  It allows you to pick & choose exactly which modules you want updated, in case you want to keep the current version of one while updating others.

`v upgrade` doesn't take any options, so it will _always_ do all of them.

Subtle, perhaps, but a very real difference.

> > > https://github.com/vlang/v/blob/master/doc/docs.md#interfaces Would be nice to have questionable design choices explained - no "implement" keyword seems like a bad design choice to me (wouldn't it be tedious to implement large interfaces when you have to manually remember all the fields you have to define?) If I knew why it was made, or maybe there was some counterpoint to what I thought, it would've made sense but without any of that people might tend to think that it's just objectively bad.
> > 
> > 
> > I'm not sure how an "implements" keyword would help. You still need to create all the pieces yourself, so adding "implements" is superfluous.
> 
> `implements` allows the compiler to warn me if I failed to implement an interface, which is helpful in multiple scenarios:
> #### 1. I mixed up types
> 
> E.g. I declared `score int`, but an interface expects `score f64`
> #### 2. I forgot something
> 
> E.g. while implementing `WebBrowser` I only defined 1317 out of 1318 methods it requires
> #### 3. I maintain multiple implementations
> 
> E.g. I have an interface `Renderer` implemented by `Opengl` on Linux and `Directx` on Windows. I add new features to `Renderer`, and implement them right away. I know I implemented `Opengl` right, because it works, but I can't be certain that I didn't make mistakes `1.` and `2.` in `Directx` because I'd have to boot to Windows and run it to know for sure
> 
> This is similar to dynamically typed languages vs statically typed: In Javascript and Python you have to _use_ some code to know that there are no type errors, while statically typed languages catch them during compilation. Without `implements` you have to _use_ a struct through an interface to know that it implements that interface, while with `implements` you know it during compilation.

So it's more a matter of a safety check than anything that helps you getting things generated.  Could be useful, but not essential.  The compiler should already be complaining if you don't implement everything you are supposed to, but still try to use it as if you did.  If that's not happening already, it's likely a bug.
 
> > > https://github.com/vlang/v/blob/master/doc/docs.md#channel-select
> > > <...> channels need libatomic <...>
> > 
> > 
> > Well, none of them are necessary for base V to work. They are needed for specific modules or features to work. Yes, some of it could be handled better, and there have been some changes to help let you know when you need to install something extra, but that hasn't been "back-ported" to all the code since it was added later. V is still a work-in-progress.
> 
> V indeed works without them, and some modules have readmes with dependencies and instructions to install them (like [sqlite](https://modules.vlang.io/sqlite.html)). But since channels are imported implicitly, maybe their dependencies should be listed with V's base dependencies, or even better, in an error message.

Yep, that would be a good addition.
 
> > > https://github.com/vlang/v/blob/master/doc/docs.md#memory-management
> > > > Remaining small percentage of objects is freed via reference counting.
> > 
> > 
> > > But on https://vlang.io it's said: "Remaining small percentage of objects is freed via GC".
> > > _I don't know which one is correct so I can't fix it._
> > 
> > 
> > They are both correct, as reference counting can be considered as a type of GC.
> 
> I assume they both should be reference counting then.

Not necessarily.  There has been some discussion about abandoning reference counting, and instead making `-gc boehm` the default.  However, that's still undecided.
 
> > > https://github.com/vlang/v/blob/master/doc/docs.md#embed_file
> > > > Instead, it will be loaded the first time your program calls `f.data()` at runtime...
> > > 
> > > 
> > > What is `f.data()`? The code snippet above doesn't have `f.data()`, nor there are any explanations later in the docs.
> > 
> > 
> > Yeah, that needs to be fixed. It should be `embedded_file.data()` to match the example code.
> 
> But there's `embedded_file.to_string()` in the example code. Should it be like in the example or should the example be changed?

I would say change the comments to make the example.
 
> > > https://github.com/vlang/v/blob/master/doc/docs.md#compile-time-reflection Not clear whether get_string and get_int are built-in functions or example functions. Add something along the lines of
> > > ```v
> > > fn get_string(data string, field_name string) string {
> > >     // ...
> > > }
> > > fn get_int(data string, field_name string) string {
> > >     // ...
> > > }
> > > ```
> > 
> > 
> > As it says at the top of that example:
> > ```v
> > // TODO: not fully implemented
> > ```
> 
> ~This should probably be highlighted a bit~ It looks like it's implemented now! I'll come up with a full working example and move `decode_User()` to a different block.

Full working examples are usually better, unless they are long.  Regular docs don't have to give full examples so much as a simple example of syntax.

If the examples are larger amounts of code, a full working example could be added under the `examples` directory, and have the doc mention that.
 
> > > https://github.com/vlang/v/blob/master/doc/docs.md#cross-platform-shell-scripts-in-v
> > > > if files.len > 0 {
> > > 
> > > 
> > > Seems redundant, the foreach loop wouldn't run with zero elements anyway
> > 
> > 
> > It's a commented out example, but it is also a safeguard against `files` not having any elements... no reason to even try the `for` loop, then.
> 
> If `files` array was empty, the loop would run zero times - `for e in []int{} { println(e) }` produces no output, so I assume `if` is safe to remove.

It is more "expensive" to set everything for the `for` loop than to do a simple integer comparison to check the length.  There's no reason to remove it.

**Comment by czkz** [comment]
> > Instead of having to type
> > ```v
> > assert Point.p.y == 4
> > ```
> > you only need
> > ```v
> > assert p.y == 4
> > ```
> 
> It might be confusing to some without it. "Why don't you have to use Point there?"

I have no idea who could be confused by this - there's just no language that uses this syntax.

C:
```c
struct Point p = {30, 4};
assert(p.y == 4);
```

C++:
```cpp
auto p = Point{30, 4};
assert(p.y == 4);
```

C#:
```cs
var p = new Point(30, 4);
Debug.Assert(p.y == 4);
```

Go:
```go
p := Point{30, 4}
if p.y != 4 { panic("assertion failed") }
```

Java:
```java
var p = new Point(30, 4);
assert p.y == 4;
```

Javascript:
```javascript
let p = new Point(30, 4);
assert(p.y == 4);
```

Nim:
```nim
let p = Point(x: 30, y: 4)
assert p.y == 4
```

Python:
```python
p = Point(30, 4)
assert p.y == 4
```

Rust:
```rust
let p = Point { x: 30, y: 4 };
assert!(p.y == 4);
```

**Comment by czkz** [comment]
> In other words, v update will (try to) update all modules if you don't list any, but if you do list some, only those will be updated. It allows you to pick & choose exactly which modules you want updated, in case you want to keep the current version of one while updating others.
> 
> v upgrade doesn't take any options, so it will always do all of them.
> 
> Subtle, perhaps, but a very real difference.

It's clear that `update` behaves differently with arguments.
But `upgrade` can only be used without arguments, and it behaves exactly like `update` without arguments.
So it looks like `upgrade` is redundant.

Looking at cmd/tools/vpm.v, `upgrade` calls `update(get_outdated())`, while `update` uses `get_installed_modules()` when passed no arguments.
Is there any difference between "updating all modules" vs "updating all modules that can be updated"?

**Comment by JalonSolov** [comment]
> > > Instead of having to type
> > > ```v
> > > assert Point.p.y == 4
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > you only need
> > > ```v
> > > assert p.y == 4
> > > ```
> > 
> > 
> > It might be confusing to some without it. "Why don't you have to use Point there?"
> 
> I have no idea who could be confused by this - there's just no language that uses this syntax.
> 
> C:
> 
> ```c
> struct Point p = {30, 4};
> assert(p.y == 4);
> ```
> 
> C++:
> 
> ```c++
> auto p = Point{30, 4};
> assert(p.y == 4);
> ```
> 
> C#:
> 
> ```cs
> var p = new Point(30, 4);
> Debug.Assert(p.y == 4);
> ```
> 
> Go:
> 
> ```go
> p := Point{30, 4}
> if p.y != 4 { panic("assertion failed") }
> ```
> 
> Java:
> 
> ```java
> var p = new Point(30, 4);
> assert p.y == 4;
> ```
> 
> Javascript:
> 
> ```js
> let p = new Point(30, 4);
> assert(p.y == 4);
> ```
> 
> Nim:
> 
> ```nim
> let p = Point(x: 30, y: 4)
> assert p.y == 4
> ```
> 
> Python:
> 
> ```python
> p = Point(30, 4)
> assert p.y == 4
> ```
> 
> Rust:
> 
> ```rust
> let p = Point { x: 30, y: 4 };
> assert!(p.y == 4);
> ```

All I can say is put it in a PR, and see what they say. ;-)

**Comment by JalonSolov** [comment]
> > In other words, v update will (try to) update all modules if you don't list any, but if you do list some, only those will be updated. It allows you to pick & choose exactly which modules you want updated, in case you want to keep the current version of one while updating others.
> > v upgrade doesn't take any options, so it will always do all of them.
> > Subtle, perhaps, but a very real difference.
> 
> It's clear that `update` behaves differently with arguments. But `upgrade` can only be used without arguments, and it behaves exactly like `update` without arguments. So it looks like `upgrade` is redundant.
> 
> Looking at cmd/tools/vpm.v, `upgrade` calls `update(get_outdated())`, while `update` uses `get_installed_modules()` when passed no arguments. Is there any difference between "updating all modules" vs "updating all modules that can be updated"?

Yes.  `v update` with no parameters will try to update every module you have installed, whether it is already up-to-date or not.  `v upgrade` will only try to update it if what you have isn't up-to-date.

**Comment by czkz** [comment]
> Yes. `v update` with no parameters will try to update every module you have installed, whether it is already up-to-date or not. `v upgrade` will only try to update it if what you have isn't up-to-date.

But since you can't update what is already up to date, because it's already up to date, `update` and `upgrade` will do the same thing.

**Comment by JalonSolov** [comment]
`upgrade` will do it quicker, because it doesn't even try on the ones that are already up-to-date.

**Comment by czkz** [comment]
> `upgrade` will do it quicker

So there's no reason to use `update` with no args then?
In other words, `update` with no args should do what `upgrade` does, and `upgrade` should be removed

**Comment by JalonSolov** [comment]
No, because some people like to see all of the modules touched/listed.  I could see the _output_ being changed.  `upgrade` only shows the names of modules it updated.  `update` could show the names with either `up-to-date` or `updated` after each.

**Comment by czkz** [comment]
I don't know then, there's nothing wrong with having both, except that it breaks the "one way to do things" principle.

**Comment by JalonSolov** [comment]
:shrug: They probably could be combined, and you could submit a PR to do so if you like.  In my mind they do slightly different things, but they are not different enough to absolutely require different commands.  Maybe keep one and add an option to it to give the behavior of the other.
